main.go

```diff
@@ -0,0 +1,295 @@
+	return &Sessions{
+		List:    make(map[*resp.Conn]*Session),
+		Backend: backend,
+	}
+}
+
+// single session
+func (ss *Sessions) NewSession() *Session {
+	return &Session{
+		Authenticated: false,
+		Username:      "",
+		Backend:       ss.Backend,
+	}
+}
+
+// create a new Session based on a new connection
+func (ss *Sessions) SessionAccept(conn *resp.Conn) bool {
+	log.Printf("Creating a new session for [%v]\n", conn.RemoteAddr)
+	ss.List[conn] = ss.NewSession()
+	return true
+}
+
+// verify a jwt token is valid
+func verifyToken(tokenStr string) (bool, string, error) {
+	token, err := jwt.Parse(tokenStr, func(token *jwt.Token) (interface{}, error) {
+		if token.Method != jwt.SigningMethodES384 {
+			return nil, fmt.Errorf("Unexpected signing method: %v", token.Header["alg"])
+		}
+		return jwtPubKey, nil
+	})
+
+	if err != nil {
+		return false, "", err
+	}
+
+	if claims, ok := token.Claims.(jwt.MapClaims); ok && token.Valid {
+		log.Printf("Token validated: %v", claims["username"])
+		return true, claims["username"].(string), nil
+	}
+
+	return false, "", nil
+}
+
+//
+// AUTH
+//
+func (ss *Sessions) Authenticate(conn *resp.Conn, args []resp.Value) bool {
+	// retreive session from connection
+	session := ss.List[conn]
+	return session.Authenticate(conn, args)
+}
+
+func (s *Session) Authenticate(conn *resp.Conn, args []resp.Value) bool {
+	if len(args) != 2 {
+		conn.WriteError(errors.New("Wrong number of arguments for 'AUTH' command"))
+		return true
+	}
+
+	valid, username, err := verifyToken(args[1].String())
+	if !valid {
+		log.Printf("Authentication failed: %v\n", err)
+		conn.WriteError(errors.New("Access denied"))
+
+		s.Authenticated = false
+		return true
+	}
+
+	s.Username = username
+	s.Authenticated = true
+
+	conn.WriteSimpleString("OK - " + s.Username)
+	return true
+}
+
+//
+// INFO
+//
+func (ss *Sessions) Info(conn *resp.Conn, args []resp.Value) bool {
+	conn.WriteString("Gateway for 0-db (zdb) compatible clients\n")
+	return true
+}
+
+//
+// EXISTS
+//
+func (ss *Sessions) Exists(conn *resp.Conn, args []resp.Value) bool {
+	// retreive session from connection
+	session := ss.List[conn]
+	return session.Exists(conn, args)
+}
+
+func (s *Session) Exists(conn *resp.Conn, args []resp.Value) bool {
+	if !s.Authenticated {
+		log.Printf("Command EXISTS for unauthenticated user [%v]\n", conn.RemoteAddr)
+		conn.WriteError(errors.New("Permission denied"))
+		return true
+	}
+
+	if len(args) != 2 {
+		conn.WriteError(errors.New("Wrong number of arguments for 'EXISTS' command"))
+		return true
+	}
+
+	reply, err := s.Backend.Server.Do("EXISTS", args[1])
+	if err != nil {
+		log.Printf("Exists failed: %v\n", err)
+		conn.WriteError(errors.New("Cannot perform query right now"))
+		return true
+	}
+
+	// forward value
+	conn.WriteInteger(int(reply.(int64)))
+
+	return true
+}
+
+//
+// SET
+//
+func (ss *Sessions) Set(conn *resp.Conn, args []resp.Value) bool {
+	// retreive session from connection
+	session := ss.List[conn]
```

Comment:
same remark you should probably handle the case where the connection is not in the map